addons/project/models/project.py

```diff
@@ -0,0 +1,832 @@
+            defaults = {'stage_id': task.stage_id.id,
+                        'name': task.name}
+            map_task_id[task.id] =  task.copy(defaults).id
+        project.write({'tasks':[(6, 0, map_task_id.values())]})
+
+    @api.multi
+    def attachment_tree_view(self):
+        self.ensure_one()
+        tasks = self.env['project.task'].search([('project_id', 'in', self.ids)])
+        domain = [
+             '|',
+             '&', ('res_model', '=', 'project.project'), ('res_id', 'in', self.ids),
+             '&', ('res_model', '=', 'project.task'), ('res_id', 'in', tasks.ids)]
+        return {
+            'name': _('Attachments'),
+            'domain': domain,
+            'res_model': 'ir.attachment',
+            'type': 'ir.actions.act_window',
+            'view_id': False,
+            'view_mode': 'kanban,tree,form',
+            'view_type': 'form',
+            'help': _('''<p class="oe_view_nocontent_create">
+                        Documents are attached to the tasks and issues of your project.</p><p>
+                        Send messages or log internal notes with attachments to link
+                        documents to your project.
+                    </p>'''),
+            'limit': 80,
+            'context': "{'default_res_model': '%s','default_res_id': %d}" % (self._name, self.id)
+        }
+
+
+class Task(models.Model):
+    _name = "project.task"
+    _description = "Task"
+    _date_name = "date_start"
+    _inherit = ['mail.thread', 'ir.needaction_mixin']
+    _order = "priority desc, sequence, date_start, name, id"
+    _mail_post_access = 'read'
+
+    def _get_default_partner(self):
+        context = dict(self.env.context)
+        if 'default_project_id' in context:
+            return self.env['project.project'].browse(context['default_project_id']).partner_id
+
+    def _get_default_stage_id(self):
+        """ Gives default stage_id """
+        return self.stage_find(self.env.context.get('default_project_id'), [('fold', '=', False)])
+
+    @api.multi
+    def _read_group_stage_ids(self, domain, read_group_order=None, access_rights_uid=None):
+        ProjectTaskType = self.env['project.task.type']
+        order = ProjectTaskType._order
+        access_rights_uid = access_rights_uid or self.env.user.id
+        if read_group_order == 'stage_id desc':
+            order = '%s desc' % order
+        if 'default_project_id' in self.env.context:
+            search_domain = ['|', ('project_ids', '=', self.env.context['default_project_id']), ('id', 'in', self.ids)]
+        else:
+            search_domain = [('id', 'in', self.ids)]
+        stage_ids = ProjectTaskType._search(search_domain, order=order, access_rights_uid=access_rights_uid)
+        stages = ProjectTaskType.sudo(access_rights_uid).browse(stage_ids)
+        result = stages.sudo(access_rights_uid).name_get()
+        # restore order of the search
+        result.sort(lambda x, y: cmp(stage_ids.index(x[0]), stage_ids.index(y[0])))
```

Comment:
1. use a keys function
2. that's an undesirably slow method of sorting as looking up list indexes is linear, collect a `{id: order}` dict and use that to look up sort order